lib/BaseViewer.js

```diff
@@ -87,52 +105,93 @@ inherits(BaseViewer, Diagram);
     bpmnDiagram = null;
   }
 
-  // done is optional
-  done = done || function() {};
-
   var self = this;
 
-  // hook in pre-parse listeners +
-  // allow xml manipulation
-  xml = this._emit('import.parse.start', { xml: xml }) || xml;
+  return new Promise(function(resolve, reject) {
 
-  this._moddle.fromXML(xml, 'bpmn:Definitions', function(err, definitions, context) {
+    // hook in pre-parse listeners +
+    // allow xml manipulation
+    xml = self._emit('import.parse.start', { xml: xml }) || xml;
 
-    // hook in post parse listeners +
-    // allow definitions manipulation
-    definitions = self._emit('import.parse.complete', {
-      error: err,
-      definitions: definitions,
-      context: context
-    }) || definitions;
+    self._moddle.fromXML(xml, 'bpmn:Definitions').then(function(result) {
+      var definitions = result.rootElement;
+      var references = result.references;
+      var parseWarnings = result.warnings;
+      var elementsById = result.elementsById;
 
-    var parseWarnings = context.warnings;
+      var context = {
+        elementsById: elementsById,
+        references: references,
+        warnings: parseWarnings
+      };
 
-    if (err) {
-      err = checkValidationError(err);
+      // hook in post parse listeners +
+      // allow definitions manipulation
+      definitions = self._emit('import.parse.complete', {
+        error: null,
+        definitions: definitions,
+        context: context
+      }) || definitions;
 
-      self._emit('import.done', { error: err, warnings: parseWarnings });
+      self.importDefinitions(definitions, bpmnDiagram).then(function(result) {
+        var allWarnings = [].concat(parseWarnings, result.warnings || []);
 
-      return done(err, parseWarnings);
-    }
+        self._emit('import.done', { error: null, warnings: allWarnings });
+
+        callIfExists(done, null, allWarnings);
+
+        return resolve({ warnings: allWarnings });
+      }).catch(function(err) {
+        var allWarnings = [].concat(parseWarnings, err.warnings || []);
 
-    self.importDefinitions(definitions, bpmnDiagram, function(err, importWarnings) {
-      var allWarnings = [].concat(parseWarnings, importWarnings || []);
+        self._emit('import.done', { error: err, warnings: allWarnings });
 
-      self._emit('import.done', { error: err, warnings: allWarnings });
+        callIfExists(done, err, allWarnings);
 
-      done(err, allWarnings);
+        return reject(addWarningsToError(err, allWarnings));
+      });
+    }).catch(function(err) {
+
+      self._emit('import.parse.complete', {
+        error: err
+      });
+
+      err = checkValidationError(err);
+
+      self._emit('import.done', { error: err, warnings: err.warnings });
+
+      callIfExists(done, err, err.warnings);
+
+      return reject(err);
     });
   });
-};
+});
+
+/**
+* The importDefinitions result.
+*
+* @typedef {Object} ImportDefinitionsResult
+*
+* @property {Array<String>} warnings
```

Comment:
Same as in the first file, it should be `Array<string>` as we are usin array of primitive values.